woocommerce/google-listings-and-ads
# Get product stats using the Merchant Center report query

### Changes proposed in this Pull Request:

Part of https://github.com/woocommerce/google-listings-and-ads/issues/2146

This PR explores a new method to gather product stats independently of the MC API ID. The past two approaches are:

1. The first approach (GLA v1.0.0) involved using product status to list all the products in the merchant center. However, this required fetching all the products in the MC, with a maximum of 250 results per request. For larger customers, this meant making multiple requests and fetching products that might not even be synced through the GLA.

2. The current approach involves using the MC API IDs created when submitting the products, stored in the metadata with the key `_wc_gla_google_ids`. Instead of fetching all the products, this approach fetches the products using the MC API ID, making it more efficient than the first approach. However, it may return multiple 404 errors if the product has been deleted in the MC and the GLA extension is not aware of it.

3. The new approach: As we're migrating to the API approach, we can't continue using the `_wc_gla_google_ids` metadata since we won't be submitting the product. Instead, Google will create the product for us, so we won't know the MC API ID. However, we do know the `offerID` that the product will have, which by default (or at least as it is now) is the WC product ID. We'll use the `offerIDs` from the "syncable" products to query the Merchant Center Reports, which allows us to retrieve a lot more results (the default page size is 1k) in one request and be independent of the MC API ID. 

### Detailed test instructions:


It is a bit difficult to test the results as you will need products in MC with different statuses. I have been able to test the "Pending", "Not Synced", and "Disapproved" and for the "Expiring" I changed the code to simulate an expiring product.  

1.  Go to the Product Feed - Overview
2. Check the stats and compare the numbers with the "develop" branch.
 


### Additional details:
- The plan is to remove the `gla_` prefix for the `offerID` for newly synced products to the MC.
- If the WC product has `_wc_gla_google_ids`, Google will need to keep the `ID` and the `offerID` saved in the metadata (confirmation on how this will work is still needed).
- I've introduced the new static function `get_offer_id` as a replacement for `get_google_product_offer_id.` Since we're not using the slug, we could simply remove it from the slug argument in` get_google_product_offer_id`, but I'm unsure if this change might affect other plugins using this function for other purposes. Maybe we could mark it as deprecated.
- In the next PR, I'll include tests for this approach.
- It is still pending to test the query with a thousand products to see the performance.

### More context:

- pcTzPl-1Rz-p2#comment-2990
- https://github.com/woocommerce/google-listings-and-ads/issues/1194


### Changelog entry

>
